post-processing/readme.qmd

```diff
@@ -0,0 +1,169 @@
+
+**Status**: Draft
+
+## Abstract
+
+tidymodels workflow objects include pre-processing and modeling steps. Pre-processors are data-oriented operations to prepare the data for the modeling function. We will be including post-processors that can do various things to model outputs (e.g. predictions). This tidyup focuses on the APIs and data structures inside of a workflow object. 
+
+
+## Motivation
+
+After modeling is complete there are occasions where model outputs, mostly predictions, might requirement modification. For example, calibration of model predictions is an effective technique for improving models. Also, some users may wish to optimize the probability threshold to call a sample an event (in a binary classification problem). 
+
+## Solution
+
+### User-facing APIs
+
+The plan is to have a small set of à la carte functions for specific operations. For example, let's look at a binary classification model with calibration and threshold optimization: 
+
+```{r}
+#| eval: false
+
+wflow_2 <- 
+  workflow() %>% 
+  add_formula(Class ~ .) %>% 
+  add_model(logistic_reg())
+
+wflow_individ <- 
+  wflow_2 %>% 
+  add_prob_calibration(cal_object) %>% 
+  add_prob_threshold(threshold = 0.7)
+```
+
+The new `add_*` functions would have `update_*()` and `remove_*()` analogs or overall functions that conduct those operations for all of the different post-processors. 
+
+Note that the order of operations matter. In the above example, the system should execute the calibration before threshold the probabilities since the calibration will change the data in a way that invalidates the new threshold. 
+
+However, the order that the operations are added to the worklflow should not matter. There are some rules that can be enacted (e.g. calibration before thresholding) but  the proposed solution is to have a prioritization scheme that can be, within reason, altered by the user. Each `add_*` function will have a `.priority` argument that has reasonable defaults (with one exception, see below).  
+
+### Current list of post-processors
+
+* Calibration of probability estimates to increase their probabilistic fidelity to the observed data
+
+```r
+# 'object' is a pre-made calibration object
+# Data Inputs: class probabilities
+# Data Outputs: class probabilities and recomputed class predictions
+add_prob_calibration(object, .priority = 1.0)
+```
+
+* For binary classification outcomes, there may be the need to optimize the probability threshold that produces hard class predictions.
+
+```r
+# Potentially tunable
+# Data Inputs: class probabilities
+# Data Outputs: class predictions
+add_prob_threshold(threshold = numeric(), .priority = 2.0)
+```
+
+* An addition of an equivocal zone where highly uncertain probability estimates where no hard class prediction is reported. 
+
+```r
+# Potentially tunable
+# Data Inputs: class probabilities
+# Data Outputs: class predictions
+add_cls_eq_zone(value = numeric(), threshold = numeric(), .priority = 3.0)
```

Comment:
In probably `probably::make_class_pred()` had a `buffer` argument that created a range of `[threshold - buffer[1], threshold + buffer[2]]` where anything inside the buffer range was marked equivocal. Maybe you could use the `buffer` arg here?

---

Maybe also name it something similar to `add_prob_threshold()` like `add_prob_threshold_buffered()` where:
- `add_prob_threshold()` always returns a factor (maybe ordered)
- `add_prob_threshold_buffered()` always returns a `<class_pred>` from probably